### PR TITLE
Week-1 - feat: implement ErrorOr and two vals from three quarters

### PR DESCRIPTION
**Pull request by DpAkyJ1A**
proof for val (f ० g ० h ० i)(x):
image

### CODE REVIEW

File: src/main/scala/karazin/scala/users/group/week1/homework/adt.scala [hunk 1]
```diff
@@ -1,53 +1,23 @@
...
-      
-      Make sure that in case of failing the method with exception
-      no exception is thrown but the case for an error is returned
-    */
-    def apply = ???
-      
-  
\ No newline at end of file
+    def apply[V](v: V): ErrorOr[V] =
+      if v == null then ErrorOr.Fail else ErrorOr.Success
```
Comment (IgorWolkov) [inline src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:23]: See my comments above for a `Fail` structure.
It's quite useful here to return `Fail` with `NullPointerException`.

Comment (IgorWolkov) [reply src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:23]: I'm not sure that `else ErrorOr.Success` part will compile. Anyway looks like you forgot about the value.

File: src/main/scala/karazin/scala/users/group/week1/homework/adt.scala [hunk 2]
```diff
@@ -1,53 +1,23 @@
...
+
+    def flatMap[Q](f: V ⇒ ErrorOr[Q]): ErrorOr[Q] =
+      this match
+        case ErrorOr.Fail   ⇒ ErrorOr.Fail
+        case ErrorOr.Success ⇒ f(v)
+
+    def map[Q](f: V ⇒ Q) : ErrorOr[Q] =
+      this match
+        case ErrorOr.Fail   ⇒ ErrorOr.Fail
+        case ErrorOr.Success ⇒ ErrorOr.Success(f(v))
```
Comment (IgorWolkov) [inline src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:19]: What if `f(v)` throws an error? Shall we return `Success` in this case? (btw. if `f(v)` throws an error then this method fails with exception, but the method must! return either `Success` or `Failure`).

File: src/main/scala/karazin/scala/users/group/week1/homework/adt.scala [hunk 3]
```diff
@@ -1,53 +1,23 @@
...
-      
-  // Companion object to define constructor
+    case Success(x: V) extends ErrorOr[V]
+
+    case Fail extends Error("Something went wrong...")
+
+    def flatMap[Q](f: V ⇒ ErrorOr[Q]): ErrorOr[Q] =
+      this match
+        case ErrorOr.Fail   ⇒ ErrorOr.Fail
+        case ErrorOr.Success ⇒ f(v)
```
Comment (IgorWolkov) [inline src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:14]: See my comment for `map`, `f(v)` could throw an error.

File: src/main/scala/karazin/scala/users/group/week1/and/three/quarters/homework/polymorphicfunctions.scala [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
-  val `(f ० g ० h ० i)(x)`: Any = ???
+  // it works!!!
+  val `I₍₂,₄₎⁴`: [`α₁`, `α₂`, `α₃`, `α₄`] ⇒ `α₁` ⇒ `α₂` ⇒ `α₃` ⇒ `α₄` ⇒ (`α₂`, `α₄`) =
+    [`α₁`, `α₂`, `α₃`, `α₄`] ⇒ (`x₁`: `α₁`) ⇒ (`x₂`: `α₂`) ⇒ (`x₃`: `α₃`) ⇒ (`x₄`: `α₄`) ⇒ (`x₂`, `x₄`);
 
-end polymorphicfunctions
+  val `(f ० g ० h ० i)(x)`: [ε, δ, γ, β, α] ⇒ (δ ⇒ ε) ⇒ ((γ ⇒ δ) ⇒ ((β ⇒ γ) ⇒ ((α ⇒ β) ⇒ (α ⇒ ε)))) =
+    [ε, δ, γ, β, α] ⇒ (f: (δ ⇒ ε)) ⇒ (g: (γ ⇒ δ)) ⇒ (h: (β ⇒ γ)) ⇒ (i: (α ⇒ β)) ⇒ (x: α) ⇒ f(g(h(i(x))));
+
+end polymorphicfunctions
```
Comment (IgorWolkov) [inline src/main/scala/karazin/scala/users/group/week1/and/three/quarters/homework/polymorphicfunctions.scala:12]: Well done!

**Comment by DpAkyJ1A** [comment]
I set up the working environment correctly (before, my project was built no matter what I wrote...)
So
I fixed some bugs but i dont know what to do with this:
image
This error i have in map and flatmap functions too.
There must be something wrong with my case Fail:
```
case Fail(e: Exception) extends ErrorOr[Throwable]
```
I read your comments in homework but didn't come to anything(

**Comment by DpAkyJ1A** [comment]
looks good now

File: src/main/scala/karazin/scala/users/group/week1/homework/adt.scala [hunk 5]
```diff
@@ -1,53 +1,29 @@
...
-      Make sure that in case of failing the method with exception
-      no exception is thrown but the case for an error is returned
-     */
-    def map = ???
-      
-  // Companion object to define constructor
+
+    case Success(x: V) extends ErrorOr[V]
+
+    case Fail(e: Throwable) extends ErrorOr[V]
```
Comment (IgorWolkov) [inline src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:11]: Much better.

File: src/main/scala/karazin/scala/users/group/week1/homework/adt.scala [hunk 6]
```diff
@@ -1,53 +1,29 @@
...
+
+    def flatMap[Q](f: V ⇒ ErrorOr[Q]): ErrorOr[Q] =
+      this match
+        case ErrorOr.Fail(e)    ⇒ ErrorOr.Fail(e)
+        case ErrorOr.Success(v) ⇒ f(v)
+
+    def map[Q](f: V ⇒ Q) : ErrorOr[Q] =
+      this match
+        case ErrorOr.Fail(e)    ⇒ ErrorOr.Fail(e)
+        case ErrorOr.Success(v) ⇒ ErrorOr.Success(f(v))
```
Comment (IgorWolkov) [inline src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:21]: See my comment above.

File: src/main/scala/karazin/scala/users/group/week1/homework/adt.scala [hunk 7]
```diff
@@ -1,53 +1,32 @@
...
+      this match
+        case ErrorOr.Fail(e)    ⇒ ErrorOr.Fail(e)
+        case ErrorOr.Success(v) ⇒ try f(v) catch {
+          case e: Throwable ⇒ ErrorOr.Fail(e)
+        }
+
+    def map[Q](f: V ⇒ Q) : ErrorOr[Q] =
+      this match
+        case ErrorOr.Fail(e)    ⇒ ErrorOr.Fail(e)
+        case ErrorOr.Success(v) ⇒ try ErrorOr.Success(f(v)) catch{
```
Comment (IgorWolkov) [inline src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:23]: Whitespace!

**Review by IgorWolkov — APPROVED** [review]
Great job!

**Pull request merged**